### DISCUSSION TITLE
Nested virtual navigation using useListNavigation

### DISCUSSION THREAD

**Discussion by AlesJiranek** [floating-ui/floating-ui#2487] [Q&A]
Hello,

I am trying to build a filterable dropdown.
I am building it basically on this example https://codesandbox.io/s/admiring-lamport-5wt3yg?file=/src/DropdownMenu.tsx with the addition of search input in the root dropdown menu. I set `virtual: true` on the `useListNavigation` hook to keep the focus on search input while navigating via keyboard Everything runs smoothly on the first level. Unfortunately, the right arrow does not open the sub-menus probably because keyboard events are caught by root button reference instead of sub-references buttons. 

Do you have an idea how this could be solved?

Thank you

**Comment by totorofly** [comment]
I encountered exactly the same situation as you. Do you have a solution yet?

**Reply by AlesJiranek** [reply]
Unfortunately not yet

**Comment by atomiks** [comment]
Supported in `v0.25.4`

**Reply by AlesJiranek** [reply]
Wow, awesome, thank you very much

**Reply by atomiks** [reply]
It should require a `FloatingTree` setup. 

Example (bit messy atm): https://github.com/floating-ui/floating-ui/blob/master/packages/react/test/visual/components/MenuVirtual.tsx

To know which item is selected (and then clicked when listening for events on the root), `useListNavigation` now accepts a `virtualItemRef` prop. It gets set when moving focus.

**Comment by totorofly** [comment]
> It should require a `FloatingTree` setup.
> 
> Example (bit messy atm): https://github.com/floating-ui/floating-ui/blob/master/packages/react/test/visual/components/MenuVirtual.tsx
> 
> To know which item is selected (and then clicked when listening for events on the root), `useListNavigation` now accepts a `virtualItemRef` prop. It gets set when moving focus.

Thank you so much for the upgrade of useListNavigation in floating-ui! Now, by simply setting a virtualItemRef as an argument for useListNavigation and upgrading to version 0.25.4, I can implement virtual navigation even in nested menus with Virtual enabled, which is truly amazing! (https://codesandbox.io/p/sandbox/jovial-mountain-89mzvx?file=/lib/components/DropdownMenu.tsx:117,8)

However, I've noticed a small issue. It seems that the type definition for UseListNavigationProps is missing the virtualItemRef, resulting in a type checking warning for
```
Component definition is missing display nameeslint(react/display-name)
(property) UseListNavigationProps.virtualItemRef?: MutableRefObject<HTMLElement | null> | undefined
```
